hazelcast/src/main/java/com/hazelcast/instance/DelegatingAddressPicker.java

```diff
@@ -0,0 +1,146 @@
+/*
+ * Copyright (c) 2008-2017, Hazelcast, Inc. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.hazelcast.instance;
+
+import com.hazelcast.config.ConfigurationException;
+import com.hazelcast.config.NetworkConfig;
+import com.hazelcast.logging.ILogger;
+import com.hazelcast.nio.Address;
+import com.hazelcast.nio.IOUtil;
+import com.hazelcast.spi.BindAddressProvider;
+
+import java.io.IOException;
+import java.net.InetAddress;
+import java.net.InetSocketAddress;
+import java.net.ServerSocket;
+import java.nio.channels.ServerSocketChannel;
+import java.util.concurrent.TimeUnit;
+
+/**
+ * Delegates picking the bind and public address for this instance
+ * to an implementation of {@link BindAddressProvider}.
+ */
+public class DelegatingAddressPicker implements AddressPicker {
+    //todo: this should be shared with DefaultAddressPicker
+    private static final int SOCKET_TIMEOUT_MILLIS = (int) TimeUnit.SECONDS.toMillis(1);
+    private static final int SOCKET_BACKLOG_LENGTH = 100;
+
+    private final ILogger logger;
+    private final BindAddressProvider bindAddressProvider;
+    private final NetworkConfig networkConfig;
+
+    private volatile InetSocketAddress bindAddress;
+    private volatile InetSocketAddress publicAddress;
+
+    private volatile ServerSocketChannel serverSocketChannel;
+
+    public DelegatingAddressPicker(BindAddressProvider bindAddressProvider, NetworkConfig networkConfig, ILogger logger) {
+        this.bindAddressProvider = bindAddressProvider;
+        this.networkConfig = networkConfig;
+        this.logger = logger;
+    }
+
+
+    @Override
+    public void pickAddress() throws Exception {
+        try {
+            bindAddress = bindAddressProvider.getBindAddress();
+            logger.info("Using bind address: " + publicAddress);
+
+            publicAddress = bindAddressProvider.getPublicAddress();
+            validatePublicAddress(publicAddress);
+            logger.info("Using public address: " + publicAddress);
+
+            serverSocketChannel = createServerSocketChannelInternal();
+
+            if (publicAddress.getPort() == 0) {
+                publicAddress = new InetSocketAddress(publicAddress.getAddress(), serverSocketChannel.socket().getLocalPort());
+            }
+        } catch (Exception e) {
+            logger.severe(e);
+            throw e;
+        }
+    }
+
+
+    private ServerSocketChannel createServerSocketChannelInternal() {
```

Comment:
This method is very similar to the one in `DefaultAddressPicker`. Can we merge then in a parent class?

Reply:
I'll see what I can do here.

Reply:
I have decided to make the refactoring in a new PR and postpone that one for 3.9.1 so we minimise the impact on the `DefaultAddressPicker` and to avoid introducing bugs just before the release.

Reply:
make sense.